### PR TITLE
Add procedural macro for deriving Data

### PR DESCRIPTION
**Pull request by ineol**
This PR adds a simple procedural macro to derive the Data trait.

It simply calls the `same` function recursively on each field. It can handle structs (with anonymous fields), and enums (with inline structs).

One thing it does _not_ do is  detect that an enum is C-like, and optimize the `same` implementation to equality.

The code is a bit rough, especially the error messages, but I wanted some feedback.

### CODE REVIEW

**Comment by ineol** [comment]
Hmmm, strangely enough, the rustfmt check passed on windows but not on mac... I did not change the offending file `src/command.rs`.

**Comment by cmyr** [comment]
Cool, this is exciting! The breakage is due to rust 1.38 just being released, which I guess has a minor change in rustfmt? I've opened #183 for that.

File: druid-derive-data/Cargo.lock [hunk 1]
```diff
@@ -0,0 +1,47 @@
+# This file is automatically @generated by Cargo.
```
Comment (cmyr) [inline druid-derive-data/Cargo.lock]: I don't think we need `Cargo.lock` to be checked in; this is usually only done for binary crates.

Comment (ineol) [reply druid-derive-data/Cargo.lock]: Oups, I did not know!

**Comment by raphlinus** [comment]
> should this be behind a feature, or do we include it always?

I'm ok making the executive decision to always depend on proc macros. Otherwise we risk fragmentation. In druid, we're going to use them for data and lenses, at least, and in Runebender we're going to use them extensively for structured font data.

**Comment by cmyr** [comment]
@ineol you should be able to rebase on origin/master and pass CI now.

**Comment by cmyr** [comment]
Cool, I'm going to merge this now, and I'll open an issue to update the docs. Thanks @ineol!

**Pull request merged**